compiler/src/pipeline.rs

```diff
@@ -0,0 +1,489 @@
+use std::{
+    fmt::Display,
+    fs,
+    path::{Path, PathBuf},
+    time::Instant,
+};
+
+use analysis::convert_analyzed_to_pil_constraints;
+use ast::{
+    analyzed::Analyzed,
+    asm_analysis::AnalysisASMFile,
+    object::PILGraph,
+    parsed::{asm::ASMProgram, PILFile},
+    DiffMonitor,
+};
+use backend::{BackendType, Proof};
+use executor::{constant_evaluator, witgen::QueryCallback};
+use log::Level;
+use mktemp::Temp;
+use number::FieldElement;
+
+use crate::verify::{write_commits_to_fs, write_constants_to_fs, write_constraints_to_fs};
+
+pub struct GeneratedWitness<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+}
+
+pub struct PilWithConstants<T: FieldElement> {
+    pub pil: Analyzed<T>,
+    pub constants: Vec<(String, Vec<T>)>,
+}
+
+pub struct CompilationResult<T: FieldElement> {
+    /// Constant columns, potentially incomplete (if success is false)
+    pub constants: Vec<(String, Vec<T>)>,
+    /// Witness columns, potentially None (if success is false)
+    pub witness: Option<Vec<(String, Vec<T>)>>,
+    /// Proof, potentially None (if success is false)
+    pub proof: Option<Proof>,
+    /// Serialized low level constraints, potentially None (if success is false)
+    pub constraints_serialization: Option<String>,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy)]
+enum Stage {
+    AsmFile,
+    AsmString,
+    Parsed,
+    Resolved,
+    AnalyzedAsm,
+    PilConstraints,
+    Graph,
+    Linked,
+    PilFile,
+    PilString,
+    AnalyzedPil,
+    OptimizedPil,
+    PilWithConstants,
+    GeneratedWitness,
+    Proof,
+}
+
+enum Artifact<T: FieldElement> {
+    AsmFile(PathBuf),
+    AsmString(Option<PathBuf>, String),
+    Parsed(Option<PathBuf>, ASMProgram<T>),
+    Resolved(ASMProgram<T>),
+    AnalyzedAsm(AnalysisASMFile<T>),
+    PilConstraints(AnalysisASMFile<T>),
+    Graph(PILGraph<T>),
+    Linked(PILFile<T>),
+    PilFile(PathBuf),
+    PilString(String),
+    AnaylyzedPil(Analyzed<T>),
+    OptimzedPil(Analyzed<T>),
+    PilWithConstants(PilWithConstants<T>),
+    GeneratedWitness(GeneratedWitness<T>),
+    Proof(CompilationResult<T>),
```

Comment:
> Actually since `Stage` and `Artifact` have the same variants, would it make sense to use only `Artifact`?

They have the same name, but the `Artifact` carries data. For example, I wouldn't know how to implement what's now `pipeline.advance_to(Stage::GeneratedWitness)`?

> Also, does it make sense to call the last one `Proof`? Since the proof is actually optional in that data structure.

Hmm, do you have a suggestion? It's kind of an attempt at creating a proof if I understand this right, or preparing files to pass to some CLI. Not sure what a good name is. Probably I should also rename `CompilationResult ` (copied this from `compiler/lib.rs`).

Reply:
Yea let's keep `Proof`